### PR TITLE
[WASMFS] rmdir and unlink syscalls

### PR DESCRIPTION
**Pull request by ethanalee**
Relevant Issue: #15041 

- Implement `rmdir` and `unlink` syscalls.
- Added to existing test `unistd/unlink.c`.

### CODE REVIEW

File: system/lib/wasmfs/file.h [hunk 1]
```diff
@@ -158,6 +158,13 @@ class Directory : public File {
       lockedInserted.setParent(file);
     }
 
+    void unlinkEntry(std::string pathName) {
+      // TODO: Unlinking the parent pointer after erasing from entries.
+      auto unlinked = getDir()->entries[pathName];
+      unlinked->locked().setParent(std::shared_ptr<File>());
+      getDir()->entries.erase(pathName);
```
Comment (tlively) [inline system/lib/wasmfs/file.h:165]: Would it make sense to hold the file lock while the directory entry is cleared as well? That would make the unlink atomic and ensure that no other part of the system can observe the half-unlinked state.

File: system/lib/wasmfs/syscalls.cpp [hunk 2]
```diff
@@ -513,4 +520,85 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
   }
   return __WASI_ERRNO_SUCCESS;
 }
+
+// This enum specifies whether rmdir or unlink is being performed.
+enum class UnlinkMode { Rmdir, Unlink };
+
+static long __unlink(char* path, UnlinkMode unlinkMode) {
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:527]: The `__` prefix is reserved for system use, and since this isn't a system call itself, it would be better to use a non-reserved name.
```suggestion
static long doUnlink(char* path, UnlinkMode unlinkMode) {
```

File: system/lib/wasmfs/syscalls.cpp [hunk 3]
```diff
@@ -513,4 +520,85 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+  auto pathParts = splitPath(path);
+
+  // Root (/) directory cannot be removed.
+  if (pathParts.size() == 1 && pathParts[0] == "/") {
+    return -EBUSY;
+  }
+
+  if (pathParts.empty() || pathParts.back() == ".") {
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:535]: I'm surprised that `.` is treated specially but `..` is not. Is that consistent with native implementations? It would probably be worth a comment mentioning that this is intentional because this is how the call is specified.

File: system/lib/wasmfs/syscalls.cpp [hunk 4]
```diff
@@ -513,4 +520,85 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
   }
   return __WASI_ERRNO_SUCCESS;
 }
+
+// This enum specifies whether rmdir or unlink is being performed.
+enum class UnlinkMode { Rmdir, Unlink };
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:525]: nice 👍

File: system/lib/wasmfs/syscalls.cpp [hunk 5]
```diff
@@ -513,4 +520,85 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+
+static long __unlink(char* path, UnlinkMode unlinkMode) {
+  auto pathParts = splitPath(path);
+
+  // Root (/) directory cannot be removed.
+  if (pathParts.size() == 1 && pathParts[0] == "/") {
+    return -EBUSY;
+  }
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:533]: It would be better to check this by pointer comparison with the root dir, since there might be other paths that resolve to the root, e.g. using symlinks or `..`.

File: system/lib/wasmfs/syscalls.cpp [hunk 6]
```diff
@@ -513,4 +520,85 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+  if (unlinkMode == UnlinkMode::Rmdir) {
+
+    if (!targetDir) {
+      return -ENOTDIR;
+    }
+
+    // The root directory or current work directory cannot be removed.
+    if (targetDir == wasmFS.getCWD()) {
+      return -EBUSY;
+    }
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:569]: Is this true? The `rmdir` man page doesn't mention this.

File: system/lib/wasmfs/syscalls.cpp [hunk 7]
```diff
@@ -513,4 +520,85 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+      return -ENOTEMPTY;
+    }
+  } else {
+    if (targetDir) {
+      return -EISDIR;
+    }
+  }
+
+  // Cannot unlink if the parent dir doesn't have write permissions.
+  if (!(lockedParentDir.mode() & 0222)) {
```
Comment (tlively) [inline system/lib/wasmfs/syscalls.cpp:582]: Are there macros we can use instead of literal `0222`? Also, it might be reasonable to leave checking the permissions as a TODO. IIRC, we don't currently check them anywhere else.

File: tests/unistd/unlink.c [hunk 8]
```diff
@@ -52,10 +52,14 @@ void setup() {
 #ifndef NO_SYMLINK
   symlink("dir-empty", "dir-empty-link");
 #endif
+#ifndef WASMFS
```
Comment (tlively) [inline tests/unistd/unlink.c:55]: It would be good to add a comment explaining why there is a difference here and when we will be able to unify the code, if ever.

File: tests/unistd/unlink.c [hunk 9]
```diff
@@ -92,10 +98,11 @@ void test() {
   err = unlink("dir-readonly");
   assert(err == -1);
 
-  // emscripten uses 'musl' what is an implementation of the standard library for Linux-based systems
+// emscripten uses 'musl' what is an implementation of the standard library
+// for Linux-based systems
```
Comment (tlively) [inline tests/unistd/unlink.c:102]: It would be nice if we could avoid making changes to unrelated comments to avoid polluting `git blame` results.

File: system/lib/wasmfs/syscalls.cpp (hunk 7 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:582]: I think we can use a macro here. I wanted to include the checking permissions to match the existing test behaviour.

File: system/lib/wasmfs/file.h (hunk 1 above)
Comment (ethanalee) [reply system/lib/wasmfs/file.h:165]: Yep, makes sense. Discussed in our meeting

File: system/lib/wasmfs/syscalls.cpp (hunk 5 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:533]: This can be part of general refactoring for path parsing. Made a todo for this.

File: system/lib/wasmfs/syscalls.cpp (hunk 3 above)
Comment (ethanalee) [reply system/lib/wasmfs/syscalls.cpp:535]: Left a TODO comment since this should realistically be handled by the path parsing logic.

File: tests/unistd/unlink.c (hunk 8 above)
Comment (ethanalee) [reply tests/unistd/unlink.c:55]: Yep, it's because `chmod` is not implemented yet.

File: system/lib/wasmfs/file.h [hunk 10]
```diff
@@ -158,6 +158,13 @@ class Directory : public File {
       lockedInserted.setParent(file);
     }
 
+    void unlinkEntry(std::string pathName) {
+      // Hold lock on target file for duration of the operation.
```
Comment (kripken) [inline system/lib/wasmfs/file.h:162]: This comment says what the code does, but not *why* it does it. I think it's important to stress that it makes the entire operation atomic, that is, the unlinking on both sides happens as if it is a single operation, and no other thread can observe state in the middle.

File: system/lib/wasmfs/file.h [hunk 11]
```diff
@@ -158,6 +158,13 @@ class Directory : public File {
...
+      auto unlinked = getDir()->entries[pathName]->locked();
+      unlinked.setParent(std::shared_ptr<File>());
```
Comment (kripken) [inline system/lib/wasmfs/file.h:164]: Does this compile?
```suggestion
      unlinked.setParent({});
```

File: system/lib/wasmfs/wasmfs.h [hunk 12]
```diff
@@ -18,6 +18,8 @@
 #include <vector>
 #include <wasi/api.h>
 
+#define WASMFS_PERM_WRITE 0222
```
Comment (kripken) [inline system/lib/wasmfs/wasmfs.h:21]: It seems odd to have this here... why is it here?

(That is - why is it here in this file, and not in the single file that currently uses it, as being here suggests it is part of the external interface. And, why is this needed at all - is there no existing define for "write permissions"?)

File: tests/test_other.py [hunk 13]
```diff
@@ -11175,6 +11175,11 @@ def test_unistd_mkdir(self):
   def test_unistd_cwd(self):
     self.do_run_in_out_file_test('wasmfs/wasmfs_chdir.c')
 
+  def test_unistd_unlink(self):
+    self.set_setting('WASMFS')
+    self.emcc_args = self.emcc_args.copy() + ['-DWASMFS', '-DNO_SYMLINK=1']
+    self.do_runf(test_file('unistd/unlink.c'), 'success')
```
Comment (kripken) [inline tests/test_other.py:11181]: Could this instead be added to the existing `test_unistd_unlink`, by adding the wasmfs decorator to it?

File: system/lib/wasmfs/wasmfs.h (hunk 12 above)
Comment (ethanalee) [reply system/lib/wasmfs/wasmfs.h:21]: Added it because the existing `stat.h` macros seemed difficult to read at a glance. Also we don't really make use of the user, group and other distinctions. It might make sense to just have a simpler macro for our use case?

File: tests/test_other.py (hunk 13 above)
Comment (kripken) [reply tests/test_other.py:11181]: I think this new test addition is no longer needed?

**Review by tlively — APPROVED** [review]

File: system/lib/wasmfs/syscalls.cpp [hunk 14]
```diff
@@ -536,6 +540,81 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
...
+    }
+  }
+
+  // Cannot unlink/rmdir if the parent dir doesn't have write permissions.
+  if (!(lockedParentDir.mode() & WASMFS_PERM_WRITE)) {
+    return -EACCES;
+  }
+
+  // Erase targetDir from lockedParentDir and also set targetDir's parent
+  // pointer to nullptr.
```
Comment (kripken) [inline system/lib/wasmfs/syscalls.cpp:606]:
```suggestion
  // Input is valid, perform the unlink.
```
1. The two mentions of `targetDir` were inaccurate/stale.
2. There is no need to mention setting the parent pointer here. That is a detail of what the unlink method does. Imagine that in the future we don't have parent pointers, or some other refactoring, then `unlinkEntry()` would do something new and correct for the new model, but this comment would become stale and incorrect.

**Review by kripken — APPROVED** [review]
lgtm % comment

**Comment by kripken** [comment]
The code size errors can be fixed by merging in latest `main`.
